mixer/adapter/dogstatsd/dogstatsd.go

```diff
@@ -0,0 +1,158 @@
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/adapter/dogstatsd/config/config.proto
+package dogstatsd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/mixer/adapter/dogstatsd/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+)
+
+type builder struct {
+	adapterConfig *config.Params
+	metricTypes   map[string]*metric.Type
+}
+type handler struct {
+	rate    float64
+	client  *statsd.Client
+	metrics map[string]*config.Params_MetricInfo
+}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(_ context.Context, env adapter.Env) (adapter.Handler, error) {
+	ac := b.adapterConfig
+
+	bufferLength := int(ac.BufferLength)
+	if bufferLength <= 0 {
+		env.Logger().Infof("Got BufferLength of '%d', defaulting to '%d'", bufferLength, statsd.OptimalPayloadSize)
+		bufferLength = statsd.OptimalPayloadSize
+	}
+	client, err := statsd.NewBuffered(ac.Address, bufferLength)
+	if err != nil {
+		env.Logger().Errorf("Unable to create dogstatsd client: %v", err)
+		return nil, err
+	}
+	client.Namespace = ac.Prefix
+	client.Tags = flattenTags(ac.GlobalTags)
+
+	// Create an empty map if tags aren't provided
+	for _, m := range ac.Metrics {
+		if m.Tags == nil {
+			m.Tags = map[string]string{}
+		}
+	}
+	return &handler{ac.SampleRate, client, ac.Metrics}, nil
+}
+
+// adapter.HandlerBuilder#SetAdapterConfig
+func (b *builder) SetAdapterConfig(cfg adapter.Config) { b.adapterConfig = cfg.(*config.Params) }
+
+// adapter.HandlerBuilder#Validate
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	ac := b.adapterConfig
+	if ac.BufferLength < 0 {
+		ce = ce.Appendf("bufferLength", "buffer length must be >= 0")
+	}
+	if ac.SampleRate < 0 || ac.SampleRate > 1 {
+		ce = ce.Appendf("sampleRate", "sampling rate must be between 0 and 1")
+	}
+	return
+}
+
+// metric.HandlerBuilder#SetMetricTypes
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+////////////////// Request-time Methods //////////////////////////
+// metric.Handler#HandleMetric
+func (h *handler) HandleMetric(_ context.Context, insts []*metric.Instance) error {
+	var result *multierror.Error
+	for _, inst := range insts {
+		if err := h.record(inst); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+func (h *handler) record(value *metric.Instance) error {
+	mname := value.Name
+	info, found := h.metrics[mname]
+
+	if !found {
+		return fmt.Errorf("no info for metric named %s", value.Name)
+	}
+	// Add dimensions to tags
+	for k, v := range value.Dimensions {
+		info.Tags[k] = fmt.Sprintf("%v", v)
+	}
+	// Take the tag map and return a tag string that dogstatsd expects
+	tags := flattenTags(info.Tags)
+
+	switch info.Type {
+	case config.COUNTER:
```

Comment:
Similarly this validation of datatype of `value` can be done during Validate call. The SetMetricType gives exactly what is the datatype of the `value` field.